### PR TITLE
oEmbed: Add support of Facebook

### LINKED ISSUE #3532: Facebook oEmbed support
### Issue Summary

When embedding facebook posts or videos using streamfield they are unsupported.

### Steps to Reproduce

1. Create a post
2. Add a streamfield embed part
3. Enter a facebook post/video eg https://www.facebook.com/20531316728/posts/10154009990506729/
4. Note nothing is show in the post

It appears support just needs to be added to this file - https://github.com/wagtail/wagtail/blob/1874b8b259e1295370b1c2122966d96668db7eec/wagtail/wagtailembeds/oembed_providers.py

Facebook oEmbed docs - https://developers.facebook.com/docs/plugins/oembed-endpoints

Note a stackoverflow answer suggests that mobiles may need to request a subtly different domain - http://stackoverflow.com/questions/5752345/does-facebook-support-oembed/33895663#33895663

### Technical details

* Python version: Python 3.5.2
* Django version: Django (1.10.5)
* Wagtail version: wagtail (1.9)
* Browser version: Firefox (52)

### PR DESCRIPTION
**Pull request by m1kola**
Fixes #3532

It looks like [Facebook's documentation](https://developers.facebook.com/docs/plugins/oembed-endpoints) a bit outdated.

The following patterns work:
```
# https://www.facebook.com/{page-name}/posts/{post-id}
https://www.facebook.com/VICE/posts/1634649483234970

# https://www.facebook.com/{username}/posts/{post-id}
https://www.facebook.com/m1kola/posts/10206566537760498

# https://www.facebook.com/photo.php?fbid={photo-id}
https://www.facebook.com/photo.php?fbid=10206529931045353&set=a.10204334520601464.1073741828.1782978439&type=3&theater

# https://www.facebook.com/notes/{username}/{note-url}/{note-id}
https://www.facebook.com/notes/shopee/makingmemories/1920792534839731/

# https://www.facebook.com/media/set?set={set-id}
https://www.facebook.com/media/set/?set=oa.1281459648537254&type=1

# https://www.facebook.com/{page-name}/videos/{video-id}/
https://www.facebook.com/1000crossroadsoflife/videos/1139992489461304/

# https://www.facebook.com/video.php?v={video-id}
https://www.facebook.com/video.php?v=1139992489461304
```

This one is not documented, but works too
```
# https://www.facebook.com/{page-name}/photos/{photo-id}
https://www.facebook.com/VICE/photos/rpp.167115176655082/1606848842681701/?type=3&theater
```

The following urls documented, but do not work:
```
# https://www.facebook.com/permalink.php?story_fbid={post-id}
https://www.facebook.com/permalink.php?story_fbid=1654663101215266&id=100000147754279

# https://www.facebook.com/{username}/videos/{video-id}/
https://www.facebook.com/natallia.ku/videos/10212852498933525/

# https://www.facebook.com/video.php?id={video-id}
https://www.facebook.com/video.php?id=1139992489461304
```

I haven't managed to find examples for the following patterns:
```
# https://www.facebook.com/{username}/activity/{activity-id}
# https://www.facebook.com/photos/{photo-id}
# https://www.facebook.com/questions/{question-id}
```

### CODE REVIEW

File: wagtail/wagtailembeds/oembed_providers.py [hunk 1]
```diff
@@ -82,6 +82,22 @@
         "^http://instagr\\.am/p/.+$",
         "^http[s]?://(?:www\\.)?instagram\\.com/p/.+$"
     ],
+    "https://www.facebook.com/plugins/video/oembed.{format}": [
+        "^https://(?:www\\.)?facebook\\.com/.+/videos/.+$",
```
Comment (BertrandBordage) [inline wagtail/wagtailembeds/oembed_providers.py]: The first `.+` may be more accurate using the non-greedy `.+?` instead.

File: wagtail/wagtailembeds/oembed_providers.py [hunk 2]
```diff
@@ -82,6 +82,22 @@
...
+        "^https://(?:www\\.)?facebook\\.com/video\\.php\\?(?:v|id)=.+$",
+    ],
+    "https://www.facebook.com/plugins/post/oembed.{format}": [
+        "^https://(?:www\\.)?facebook\\.com/.+/(?:posts|activity)/.+$",
```
Comment (BertrandBordage) [inline wagtail/wagtailembeds/oembed_providers.py]: The first `.+` may be more accurate using the non-greedy `.+?` instead.

File: wagtail/wagtailembeds/oembed_providers.py [hunk 3]
```diff
@@ -82,6 +82,22 @@
...
+        "^https://(?:www\\.)?facebook\\.com/photo\\.php\\?fbid=.+$",
+        "^https://(?:www\\.)?facebook\\.com/(?:photos|questions)/.+$",
+        "^https://(?:www\\.)?facebook\\.com/permalink\\.php\\?story_fbid=.+$",
+        "^https://(?:www\\.)?facebook\\.com/media/set/?\\?set=.+$",
+        "^https://(?:www\\.)?facebook\\.com/notes/.+/.+/.+$",
```
Comment (BertrandBordage) [inline wagtail/wagtailembeds/oembed_providers.py]: The first two `.+` may be more accurate using the non-greedy `.+?` instead.

File: wagtail/wagtailembeds/oembed_providers.py [hunk 4]
```diff
@@ -82,6 +82,22 @@
...
+
+        # At the moment, not documented on https://developers.facebook.com/docs/plugins/oembed-endpoints
+        # Works for posts with a single photo
+        "^https://(?:www\\.)?facebook\\.com/.+/photos/.+$",
```
Comment (BertrandBordage) [inline wagtail/wagtailembeds/oembed_providers.py]: The first `.+` may be more accurate using the non-greedy `.+?` instead.

**Comment by m1kola** [comment]
> I made some non-greedy suggestions that may not be accurate, depending on what URLs are supposed to contain.

@BertrandBordage all makes sense. I've updated the PR. I was going to make these changes after similar comment in #3548, but forgot :)

I'm not sure that our regexes cover all possible urls. I haven't managed to find a documentation which characters URLs can contain. I think we should go with the current patterns for now and fix issues if someone report any.

**Comment by m1kola** [comment]
> Also: do you know why no regexp in this file uses the raw string prefix and use double escaping instead?

I'm not sure, but it seems to me that they were copied from https://github.com/panzi/oembedendpoints/blob/bfc1736fa1778fce714c810467d0a6656ce396f5/endpoints-regexp.json

**Comment by ahayzen** [comment]
FWIW, this patch fixes the issue I originally reported, I hope it gets merged into the next release :-) thanks @m1kola

**Comment by gasman** [comment]
Merged in 0611a1b2cdc9ac232ddb22c0781215198d885909

**Pull request closed without merging**